### ISSUE TITLE
Decoding validation too strict

### ISSUE THREAD

**Issue by macfol** [InteractiveAdvertisingBureau/iabtcf-java#129]
I use your library to encode a tc string and it works well but when i try to decode the generated string it throws error with below message:
`com.iabtcf.exceptions.InvalidRangeFieldException: start vendor id (32797) is greater than endVendorId (16398)`

Code to repro the error:

```java
import com.iabtcf.decoder.TCString;
import com.iabtcf.encoder.PublisherRestrictionEntry;
import com.iabtcf.encoder.TCStringEncoder;
import com.iabtcf.utils.BitSetIntIterable;
import com.iabtcf.v2.RestrictionType;
import org.junit.jupiter.api.Test;

import java.time.Instant;
import java.util.Collection;
import java.util.List;
import java.util.Set;

import static org.assertj.core.api.Assertions.assertThat;

public class Playground {

    @Test
    void test() {
        final String encoded = TCStringEncoder.newBuilder()
                .version(2)
                .created(Instant.now())
                .lastUpdated(Instant.now())
                .tcfPolicyVersion(2)
                .cmpId(1)
                .isServiceSpecific(true)
                .useNonStandardStacks(false)
                .purposeOneTreatment(false)
                .publisherCC("DE")
                .cmpVersion(1)
                .consentLanguage("DE")
                .vendorListVersion(108)
                .addVendorConsent(BitSetIntIterable.from(700))
                .addVendorLegitimateInterest(BitSetIntIterable.from(700))
                .addPurposesConsent(BitSetIntIterable.from(Set.of(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)))
                .addPurposesLITransparency(BitSetIntIterable.from(Set.of(2, 3, 4, 5, 6, 7, 8, 9, 10)))
                .addPublisherRestrictionEntry(createRestrictions())
                .encode();

        final TCString decoded = TCString.decode(encoded);
        assertThat(decoded).isNotNull();
    }

    private Collection<PublisherRestrictionEntry> createRestrictions() {
        return List.of(
                PublisherRestrictionEntry.newBuilder()
                        .purposeId(2)
                        .restrictionType(RestrictionType.REQUIRE_CONSENT)
                        .addVendor(BitSetIntIterable.from(Set.of(11, 21, 25, 28, 34, 50, 60, 68, 79, 85, 195, 253, 580, 804, 807, 814)))
                        .build(),

                PublisherRestrictionEntry.newBuilder()
                        .purposeId(3)
                        .restrictionType(RestrictionType.REQUIRE_CONSENT)
                        .addVendor(BitSetIntIterable.from(Set.of(76, 92, 469, 580, 807)))
                        .build(),

                PublisherRestrictionEntry.newBuilder()
                        .purposeId(4)
                        .restrictionType(RestrictionType.REQUIRE_CONSENT)
                        .addVendor(BitSetIntIterable.from(Set.of(76, 469, 580, 807)))
                        .build(),

                PublisherRestrictionEntry.newBuilder()
                        .purposeId(5)
                        .restrictionType(RestrictionType.REQUIRE_CONSENT)
                        .addVendor(BitSetIntIterable.from(Set.of(76, 92, 667, 755)))
                        .build(),

                PublisherRestrictionEntry.newBuilder()
                        .purposeId(6)
                        .restrictionType(RestrictionType.REQUIRE_CONSENT)
                        .addVendor(BitSetIntIterable.from(Set.of(76, 755)))
                        .build(),

                PublisherRestrictionEntry.newBuilder()
                        .purposeId(7)
                        .restrictionType(RestrictionType.REQUIRE_CONSENT)
                        .addVendor(BitSetIntIterable.from(Set.of(11, 21, 25, 28, 34, 50, 60, 62, 68, 79, 85, 92, 95, 98, 152, 195, 253, 278, 580, 804, 807, 814)))
                        .build(),

                PublisherRestrictionEntry.newBuilder()
                        .purposeId(8)
                        .restrictionType(RestrictionType.REQUIRE_CONSENT)
                        .addVendor(BitSetIntIterable.from(Set.of(11, 25, 34, 76, 85, 92, 95, 98, 580, 807)))
                        .build(),

                PublisherRestrictionEntry.newBuilder()
                        .purposeId(9)
                        .restrictionType(RestrictionType.REQUIRE_CONSENT)
                        .addVendor(BitSetIntIterable.from(Set.of(11, 25, 28, 34, 60, 68, 85, 92, 95, 98, 580, 807)))
                        .build(),

                PublisherRestrictionEntry.newBuilder()
                        .purposeId(10)
                        .restrictionType(RestrictionType.REQUIRE_CONSENT)
                        .addVendor(BitSetIntIterable.from(Set.of(11, 21, 25, 28, 34, 50, 60, 68, 79, 82, 85, 95, 98, 253, 278, 580, 804, 807, 814)))
                        .build()
        );
    }
}
```

**Comment by macfol** [comment]
IMO the decoding validation should be less strict: If encoding passed why decoding should fail?

**Comment by laktech** [comment]
the above referenced code seems to pass fine for me.

**Comment by macfol** [comment]
Well it definitely doesn't work for me. Which JDK r u using?
IMO decoder should allow to decode the string, especially when it was created using provided encoder. 

![image](https://user-images.githubusercontent.com/12062146/134320340-f38161f6-dc98-4d11-9273-a9bd2171941e.png)

**Comment by macfol** [comment]
` what caused the InvalidRangeFieldException`
Well this part fails:   `final TCString decoded = TCString.decode(encoded);`

**Comment by macfol** [comment]
@laktech i noticed this issue when we introduced publisher restrictions. For everything else you are using BitSetIterable which olds integers in natural order but for publisher restrictions u use regular java collection and the order is as I which. I noticed that error occurs simply but swithicng the order of the restrictions I create. So far i kept reversed order (purpose 10,9,8 etc)
But with latest GVL change one of the vendor dropped and suddenly the mentioned error started to occur. This seems really unstable... 😞

**Comment by laktech** [comment]
That's not how publisher restrictions are handled. They vendor ids are maintained in a set data structure where order does not matter.

I'm unable to reproduce your issue with the code snippet provided. I tested it under JDK1.8 and the latest JDK 17

**Comment by macfol** [comment]
I have created a small project to test some decoding scenarios and it clearly fails.
Failed build [here](https://github.com/macfol/IAB_POC/runs/3720362449)
Project [here](https://github.com/macfol/IAB_POC)
![image](https://user-images.githubusercontent.com/12062146/134910126-75e4617a-b53c-4a2a-9f5d-95527a145aa1.png)

**Comment by macfol** [comment]
Could you check that out @laktech ?

**Comment by macfol** [comment]
Also I have noticed that in master branch you have defined version 2.0.9 and in maven repo the latest available is 2.0.8

**Comment by laktech** [comment]
thanks, looks like you're hitting a bug in 2.0.8 which was fixed in 12e6d619b7eb3c1c46ee9a58155079503b86bfdc on master but not yer released. i've requested that we release 2.0.9 in the next few days which should resolve your issue.

if possible, try to use the latest master until 2.0.9 is released.

**Comment by macfol** [comment]
Thanks!
